gobblin-data-management/src/main/java/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriter.java

```diff
@@ -155,9 +158,39 @@ private void setPathPermission(Path path, OwnerAndPermission ownerAndPermission)
   private void setRecursivePermission(Path path, OwnerAndPermission ownerAndPermission) throws IOException {
     List<FileStatus> files = FileListUtils.listPathsRecursively(fs, path, FileListUtils.NO_OP_PATH_FILTER);
 
+    // Set permissions bottom up. Permissions are set to files first and then directories
+    Collections.reverse(files);
+
     for (FileStatus file : files) {
-      setPathPermission(file.getPath(), ownerAndPermission);
+      setPathPermission(file.getPath(), addExecutePermissionsIfRequired(file, ownerAndPermission));
+    }
+  }
+
+  /**
+   * The method makes sure it always grants execute permissions for an owner if the <code>file</code> passed is a
+   * directory. The publisher needs it to publish it to the final directory and list files under this directory.
+   */
+  private OwnerAndPermission addExecutePermissionsIfRequired(FileStatus file, OwnerAndPermission ownerAndPermission) {
+
+    if (!file.isDir()) {
+      return ownerAndPermission;
+    }
+    FsAction newOwnerAction = ownerAndPermission.getFsPermission().getUserAction();
+
+    if (ownerAndPermission.getFsPermission().getUserAction() == FsAction.READ) {
```

Comment:
use a `switch...case` statement instead, and throw `IllegalStateException` if it is anything but `READ`, `WRITE`, `READ_WRITE`.